aionetbox/api.py

```
@@ -0,0 +1,302 @@
+
+class NetboxApiOperation():
+
+    def __init__(self, tag, operation, config, client):
+        self.client = client
+        self.config = config
+        self.rest_config = self.config.get('rest', {})
+        self.operation = operation
+        self.tag = tag
+
+    async def request(self, **kwargs):
+
+        path, body, query = self.parse_params(kwargs)
+
+        if self.operation_method == 'list':
+            # netbox has terrible swagger spec generation and leaves out custom fields (cf)
+            keys_used = set(list(body.keys()) + list(query.keys()) + list(path.keys()))
+            remaining_vars = set(kwargs.keys()) - keys_used
+
+            for kw in remaining_vars:
+                if kw.startswith('cf_'):
+                    query[kw] = kwargs.get(kw)
+
+        url = self.build_url(self.rest_config.get('url')).format(**path)
+
+        resp = await self.client.request(
+            method=self.rest_config.get('method'),
+            url=url,
+            query_params=query,
+            body=body
+        )
+
+        resp.raise_for_status()
+        data = await resp.json()
+        return NetboxResponseObject.from_response(
+            data=data,
+            **self.config.get('responses', {}).get(str(resp.status), {}).get('schema')
+        )
+
+    def build_url(self, url):
+        return '{}{}{}'.format(self.client.host, self.client.config.get('_orig', {}).get('basePath'), url)
+
+    def parse_params(self, params):
+        qb = {
+            'body': {},
+            'query': {},
+            'path': {},
+        }
+
+        spec_parameters = self.config.get('parameters', []) + self.rest_config.get('params', [])
+        for sp in spec_parameters:
+            if sp['name'] not in params and sp['required']:
+                raise MissingRequiredParam('{} is a required parameter'.format(sp['name']))
+
+            if sp['name'] not in params:
+                continue
+
+            qb[sp['in']][sp['name']] = params.get(sp['name'])
```

not sure what the spec parameters look like and what the data guarantees are there, but if `in` or `name` is ever missing, this will raise an error. maybe thats okay, just double checking

---

in and name would have thrown an error during spec parsing. It's probably safe to leave it and if it fails I'd rather have a key error than silently continuing